x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -71,34 +77,131 @@ export class EnterpriseSearchRequestHandler {
+          return this.handleInvalidDataError(response, url, json);
         }
+
+        // Pass successful responses back to the front-end
+        return response.custom({ statusCode: status, body: json });
       } catch (e) {
-        const errorMessage = `Error connecting to Enterprise Search: ${e?.message || e.toString()}`;
+        // Catch connection/auth errors
+        return this.handleConnectionError(response, e);
+      }
+    };
+  }
 
-        this.log.error(errorMessage);
-        if (e instanceof Error) this.log.debug(e.stack as string);
+  /**
+   * Attempt to grab a usable error body from Enterprise Search - this isn't
+   * always possible because some of our internal endpoints send back blank
+   * bodies, and sometimes the server sends back Ruby on Rails error pages
+   */
+  async getErrorResponseBody(apiResponse: Response) {
+    const { statusText } = apiResponse;
+    const contentType = apiResponse.headers.get('content-type') || '';
 
-        return response.customError({ statusCode: 502, body: errorMessage });
-      }
+    // Default response
+    let body: IErrorResponse = {
+      message: statusText,
+      attributes: { errors: [statusText] },
     };
+
+    try {
+      if (contentType.includes('application/json')) {
+        // Try parsing body as JSON
+        const json = await apiResponse.json();
+
+        // Some of our internal endpoints return either an `error` or `errors` key,
+        // which can both return either a string or array of strings ¯\_(ツ)_/¯
+        const errors = json.error || json.errors || [statusText];
+        body = {
+          message: errors.toString(),
+          attributes: { errors: Array.isArray(errors) ? errors : [errors] },
+        };
+      } else {
+        // Try parsing body as text/html
+        const text = await apiResponse.text();
+        if (text) {
+          body = {
+            message: text,
+            attributes: { errors: [text] },
+          };
+        }
+      }
+    } catch {
+      // Fail silently
+    }
+
+    return body;
+  }
+
+  /**
+   * Error response helpers
+   */
+
+  async handleClientError(response: KibanaResponseFactory, apiResponse: Response) {
+    const { status } = apiResponse;
+    const body = await this.getErrorResponseBody(apiResponse);
+
+    return response.customError({ statusCode: status, body });
+  }
+
+  async handleServerError(response: KibanaResponseFactory, apiResponse: Response, url: string) {
+    const { status } = apiResponse;
+    const { message } = await this.getErrorResponseBody(apiResponse);
+
+    // Don't expose server errors to the front-end, as they may contain sensitive stack traces
+    const errorMessage =
+      'Enterprise Search encountered an internal server error. Please contact your system administrator if the problem persists.';
+
+    this.log.error(`Enterprise Search Server Error ${status} at <${url}>: ${message}`);
+    return response.customError({ statusCode: 502, body: errorMessage });
```

Comment:
I'm a bit torn on making this a 502 error. Doing so means any 5xx state returned by an endpoint will result in the Error Connecting UI/state replacing whatever previous UI was on the page.

I'm not sure if that's too jarring compared to keeping the same page but showing an 'Server Error' flash message. On the other hand, the Error Connecting UI does have some troubleshooting steps & a guide that _could_ help folks solve intermittent 500 issues 🤷 

Any strong feelings one way or another from y'all?

Reply:
What if we did the flash message with a link to the error connecting page?

Reply:
I'd have to either have to change the 502 code (probably to just a generic 500?) or [change our 502 http interceptor itself](https://github.com/elastic/kibana/blob/master/x-pack/plugins/enterprise_search/public/applications/shared/http/http_logic.ts#L64-L70) since that logic looks for all 502 statuses and sets the top-level app state to error connecting.

I don't feel super strongly one way or another honestly - happy to put this to a vote. If no one else feels super strongly, we could also leave it as-is and change it later if we encounter a bunch of 500s and we find this UX inferior the road? 🤷